.github/workflows/requirements-3-12.txt

```diff
@@ -0,0 +1,97 @@
+#
+# This file is autogenerated by pip-compile with Python 3.12
```

Comment:
This documentation is good(ness). However, I'm not impressed by the apparent need to generate separate files for different python versions: Do I understand this right? If so, this seems to add an undue maintenance burden: Which requirements.txt file is master? How to generate the different files? How many are required? One for each python version? In other words, will this break once CI upgrades python? How to handle any errors the hashes may generate?

Reply:
thanks for the comments, and Yes, you do....

Once ANY hashes are used in the requirements.txt, all dependencies must be qualified by hashes, including the transitive dependencies. So those lists must be complete

Currently the 'unix' build (running on mac) uses Python 3.12 ie current, whilst the Windows (2022) environment is using Python 3.9. There are some dependencies we use that have slightly different dependencies themselves when being used on python 3.9 or 3.12

So this means two different files, which are likely (not guaranteed) to be correct for minor versions, and most probably often major versions

Currently there is no concept introduced of a 'master' dependency. If we wanted this, the clearest change would probably be to build using a specific version of python, and to ensure this is installed within the relevant build script - whether it's 3.9, 3.12, or something else since they can, and will vary

Whilst the python dependence could affect our direct dependencies, currently it seems it's only our transitive dependencies, so we could provide a script which builds the hash-inclusive requirements.txt from a less precise list. However we can't include this as part of the build process without going against the spirit of the OSSF checks, since the intent is to ensure there is always manual review when any change, and that the subsequent builds are repeatable. Any script would also need access to the relevant python version. It's not dissimilar to what I did to develop the change, and I totally agree it should form part of the documentation.

If CI upgrades python, yes, there is a chance of breakage. As above it's most likely on major version changes, and we know there is such a change between 3.9 & 3.12 since we have two distinct files. We could look into what version guarantees the CI images offer, but again the way to reduce dependence (and ensure more repeatability) would be to specify the version of all build tools being used, including python.

If the dependency chain is not correct, the python will fail, probably in the pip install step.

I think the best way to
 - reduce the changes of such changes breaking us
 - remove the need for two distinct requirements files (I hope, assuming there's no platform dependence...)
 
 is to use the [setup-python](https://github.com/actions/setup-python) github action, and explicitly specify python 3.12. **I could make a change to the build scripts for this and refactor?**

Reply:
Scripts are changed to use python 3.12